src/plots/cartesian/select.js

```diff
@@ -0,0 +1,102 @@
+*/
+
+
+'use strict';
+var polygon = require('../../lib/polygon');
+var axes = require('./axes');
+var filteredPolygon = polygon.filter;
+var polygonTester = polygon.tester;
+var BENDPX = 1.5; // max pixels off straight before a line counts as bent
+
+function getAxId(ax) { return ax._id; }
+
+module.exports = function prepSelect(e, startX, startY, dragOptions, mode) {
+    var plot = dragOptions.plotinfo.plot,
+        dragBBox = dragOptions.element.getBoundingClientRect(),
+        x0 = startX - dragBBox.left,
+        y0 = startY - dragBBox.top,
+        x1 = x0,
+        y1 = y0,
+        path0 = 'M' + x0 + ',' + y0,
+        pw = dragOptions.xaxes[0]._length,
+        ph = dragOptions.yaxes[0]._length,
+        xAxisIds = dragOptions.xaxes.map(getAxId),
+        yAxisIds = dragOptions.yaxes.map(getAxId);
+
+    if(mode === 'lasso') {
+        var pts = filteredPolygon([[x0, y0]], BENDPX);
+    }
+
+    var outlines = plot.selectAll('path.select-outline').data([1,2]);
+
+    outlines.enter()
+        .append('path')
+        .attr('class', function(d) { return 'select-outline select-outline-' + d; })
+        .attr('d', path0 + 'Z');
+
+    // find the traces to search for selection points
+    var searchTraces = [],
+        gd = dragOptions.gd,
+        i,
+        cd,
+        trace,
+        searchInfo,
+        selection = [];
+    for(i = 0; i < gd.calcdata.length; i++) {
+        cd = gd.calcdata[i];
+        trace = cd[0].trace;
+        if(!trace._module || !trace._module.selectPoints) continue;
+
+        if(xAxisIds.indexOf(trace.xaxis) === -1) continue;
+        if(yAxisIds.indexOf(trace.yaxis) === -1) continue;
+
+        searchTraces.push({
+            selectPoints: trace._module.selectPoints,
+            cd: cd,
+            xaxis: axes.getFromId(gd, trace.xaxis),
+            yaxis: axes.getFromId(gd, trace.yaxis)
+        });
+    }
+
+    dragOptions.moveFn = function(dx0, dy0) {
+        var poly;
+        x1 = Math.max(0, Math.min(pw, dx0 + x0));
+        y1 = Math.max(0, Math.min(ph, dy0 + y0));
+
+        if(mode === 'select') {
+            poly = polygonTester([[x0, y0], [x0, y1], [x1, y1], [x1, y0]]);
+            outlines.attr('d', path0 + 'H' + x1 + 'V' + y1 + 'H' + x0 + 'Z');
+        }
+        else if(mode === 'lasso') {
+            pts.addPt([x1, y1]);
+            poly = polygonTester(pts.filtered);
+            outlines.attr('d', 'M' + pts.filtered.join('L') + 'Z');
+        }
+
+        selection = [];
+        for(i = 0; i < searchTraces.length; i++) {
+            searchInfo = searchTraces[i];
+            [].push.apply(selection, searchInfo.selectPoints(searchInfo, poly));
+        }
+        dragOptions.gd.emit('plotly_selecting', {points: selection});
+    };
+
+    dragOptions.doneFn = function(dragged, numclicks) {
+        if(!dragged && numclicks === 2) dragOptions.doubleclick();
+        else {
+            dragOptions.gd.emit('plotly_selected', {points: selection});
```

Comment:
I also vote for keeping the selection on mouseup.

In order to have a ‘transient’ (i.e. persistent) brush, you also want to keep the selection after mousedown(s). Transient brushes are useful for selecting non-contiguous regions — here is a silly example:

![persist](https://cloud.githubusercontent.com/assets/1365941/12098692/6f35c242-b2e9-11e5-986b-94ca85169aeb.gif)


It’d be neat to have a simple way for users to flip between transient/non-transient modes (maybe with a drop-down menu when clicking on the modebar?).

And it seems as though we could reserve a double-click for clearing selections when in a “selection mode”, right? That seems like an intuitive analog given the double-click behavior for zooming.